File: 2023/AdventOfCode/Utils/Grid.fs
module Grid

type Grid<'a>(width: int, height:int, init: (int -> int -> 'a)) = 
    member private this.array = [| for x in 0..(width-1) -> [|for y in 0..(height-1) -> (init x y) |] |]

    member this.Item
        with get (x, y) = this.array[x][y]
        and set (x, y) value = this.array[x][y] <- value

    member this.Width = width
    member this.Height = height

    member this.IsValidCoord(x: int, y:int) =
        x >= 0 && y >= 0 && x < this.Width && y < this.Height

    new(width, height, init) = Grid(width, height, init)

File: 2023/TotallyFSharpHelperLibrary/Grid.cs
namespace TotallyFSharpHelperLibrary;
internal class Grid<T> {
    private readonly T[,] _array;

    public int Width { get; }
    public int Height { get; }

    public T this[int x, int y] {
        get => _array[x, y];
        set => _array[x, y] = value;
    }

    public Grid(int width, int height) {
        _array = new T[width, height];
        Width = width;
        Height = height;
    }

    public Grid(int width, int height, Func<int, int, T> initFunction) : this(width, height) {
        for (int x = 0; x < Width; x++) {
            for (int y = 0; y < Height; y++) {
                _array[x, y] = initFunction(x, y);
            }
        }
    }

    public bool IsValidCoord(int x, int y) {
        return x >= 0 && y >= 0 && x < Width && y < Height;
    }

}
